### PR TITLE
Basic twoword classifier

### PR DESCRIPTION
**Pull request by evvahu**
This branch includes a basic classifier that takes two words as input.

- basic_twoword_classifier.py: includes the class BasicTwoWordClassifier. It inherits the base class Module and overides the method "forward". The simple **concat** method is used to concatenate the two word vectors

- test_basic_twoword_classifier.py: sets up a test environment to test the class above. However, it does not really contain a proper test. It only prints out the network to check whether the dimensions and layers are correct.

### CODE REVIEW

File: scripts/basic_twoword_classifier.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        """
+        this function takes two words, concatenates them and applies a non-linear matrix transformation (hidden layer)
+        Its output is then fed to an output layer. Then it returns the concatenated and transformed vectors.
+        :param word1: the first word of size batch_size x embedding size
+        :param word2: the first word of size batch_size x embedding size
+        :return: the transformed vectors after output layer
+        """
+        word_composed = comp_functions.concat(word1, word2, axis=1)
+        x = F.relu(self.hidden_layer(word_composed))
+        return F.relu(self.output_layer(x))
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py]: return F.relu(self.output_layer(x))
This now applies a relu afterwards on top of your output layer, but the output layer is already the layer that spits out the label scores. So either you would put a softmax (multi class) or sigmoid (binary class) on top but in this case we said we would like to return "raw" class scores because there are some pytorch loss functions that do the softmax normalization and loss computation in one go. To make a long story short: remove the second relu ^^

File: scripts/basic_twoword_classifier.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import torch.nn as nn
+import torch.nn.functional as F
+import scripts.composition_functions as comp_functions
+
+
+class BasicTwoWordClassfier(nn.Module):
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py]: return F.relu(self.output_layer(x))
Typo --> change to Classifier

File: tests/test_basic_twoword_classifier.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        self.labels = 2
+
+    def test_forward(self):
+        """
+        tests the classifier implemented in BasicTwoWordClassifier and the overridden method "forward"
+        prints the classifier to check whether the right number of layers exist and the right dimensions are included
+        """
+        classifier = BasicTwoWordClassfier(input_dim=self.inputdim, hidden_dim=self.hiddendim, label_nr=self.labels)
+        classifier.forward(self.word1, self.word2)
+        print(classifier)
```
Comment (Blubberli) [inline tests/test_basic_twoword_classifier.py]: I would say: batter look at the result that you get from "forward". It should be a matrix [batchsize, label_size] in your case, because batch_size is 1 a vector of shape [1, 2]. The test does not really test something because you manually have to look at the print statement and see whether the shape is correct. instead you can store the result from forward in a variable and call the np.assert.. method to ensure that the shape of your result has the shape you want. I also think that this test is enough for now. we can add more tests later

File: tests/test_basic_twoword_classifier.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    this class tests the BasicTwoWordClassifier
+    This test suite can be ran with:
+        python -m unittest -q tests.BasicTwoWordClassifierTest
+    """
+
+    def setUp(self):
+        self.word1 = torch.from_numpy(np.array([[1, 0, 0]], dtype='float32'))
+        self.word2 = torch.from_numpy(np.array([[0, 1, 0]], dtype='float32'))
+        self.inputdim = 6
```
Comment (Blubberli) [inline tests/test_basic_twoword_classifier.py]: for consistency: let's stick to pep conventions and use input_dim , hidden_dim ... just like you did in the classifier.

File: scripts/basic_twoword_classifier.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    :param input_dim: the dimension of the input vector where only embedding size (2 times the size of the embedding of
+    a single word vector) is needed, batch size is implicit
+    :param hidden_dim : the dimension of the hidden layer, batch size is implicit
+    :param label_nr : the dimension of the output layer, i.e. the number of labels
+    """
+
+    def __init__(self, input_dim, hidden_dim, label_nr):
+        super(BasicTwoWordClassfier, self).__init__()
+        self.hidden_layer = nn.Linear(input_dim, hidden_dim)
+        self.output_layer = nn.Linear(hidden_dim, label_nr)
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py]: self.hidden_layer = nn.Linear(input_dim, hidden_dim)
        self.output_layer = nn.Linear(hidden_dim, label_nr)

it is not necessary for the code to work but because we are defining a real class now I would prefer class variables to be protected. this is done with self._hidden_layer .. you can then define them as class properties with the following:
```
@property
def hidden_layer(self):
     return self._hidden_layer

```

Comment (Blubberli) [reply scripts/basic_twoword_classifier.py]: It is a bit confusing how to do object oriented programming with Python and a lot of things will just work one or the other way because python. but it is more of a conventional thing to make class instance variables "protected".
I found this short summary useful:
https://www.tutorialsteacher.com/python/private-and-protected-access-modifiers-in-python

**Review by Blubberli — CHANGES_REQUESTED** [review]
Yeah nice first PR! Good work, only some small changes. I think in general your code looks very clean and you have a good documentation 👍

File: scripts/basic_twoword_classifier.py [hunk 6]
```diff
@@ -26,5 +34,5 @@ def forward(self, word1, word2):
         :return: the transformed vectors after output layer
         """
         word_composed = comp_functions.concat(word1, word2, axis=1)
-        x = F.relu(self.hidden_layer(word_composed))
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py]: this line should not be deleted, now the classifier has no hidden layer

File: tests/test_basic_twoword_classifier.py [hunk 7]
```diff
@@ -14,15 +14,16 @@ class BasicTwoWordClassifierTest(unittest.TestCase):
     def setUp(self):
         self.word1 = torch.from_numpy(np.array([[1, 0, 0]], dtype='float32'))
         self.word2 = torch.from_numpy(np.array([[0, 1, 0]], dtype='float32'))
-        self.inputdim = 6
-        self.hiddendim = 6
+        self.input_dim = 6
+        self.hidden_dim = 6
```
Comment (Blubberli) [inline tests/test_basic_twoword_classifier.py:18]: if you change hidden_dim to a smaller value the test below will fail, because the hidden layer has been deleted and the shapes do not match anymore.

File: tests/test_basic_twoword_classifier.py [hunk 8]
```diff
@@ -14,15 +14,16 @@ class BasicTwoWordClassifierTest(unittest.TestCase):
...
     def test_forward(self):
         """
         tests the classifier implemented in BasicTwoWordClassifier and the overridden method "forward"
-        prints the classifier to check whether the right number of layers exist and the right dimensions are included
+        checks whether the output layer is of the right size
         """
-        classifier = BasicTwoWordClassfier(input_dim=self.inputdim, hidden_dim=self.hiddendim, label_nr=self.labels)
-        classifier.forward(self.word1, self.word2)
-        print(classifier)
+        self.expected_size = torch.tensor(np.array([[0,1]])).shape
```
Comment (Blubberli) [inline tests/test_basic_twoword_classifier.py]: class variable only make sense if they are defined in the setup method and will eventually be reused by another test or if they are very general. the result of a test will only be used in one specific test in most cases so I would not define it as a class variable here.

File: scripts/basic_twoword_classifier.py [hunk 9]
```diff
@@ -13,9 +13,17 @@ class BasicTwoWordClassfier(nn.Module):
...
 
     def __init__(self, input_dim, hidden_dim, label_nr):
-        super(BasicTwoWordClassfier, self).__init__()
-        self.hidden_layer = nn.Linear(input_dim, hidden_dim)
-        self.output_layer = nn.Linear(hidden_dim, label_nr)
+        super(BasicTwoWordClassifier, self).__init__()
+        self._hidden_layer = nn.Linear(input_dim, hidden_dim)
+        self._output_layer = nn.Linear(hidden_dim, label_nr)
+
+    @property
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py]: put all getters or setters below all functions

**Review by Blubberli — CHANGES_REQUESTED** [review]
fast fertig ;)

File: scripts/basic_twoword_classifier.py (hunk 6 above)
Comment (evvahu) [reply scripts/basic_twoword_classifier.py]: ouch, sorry

File: tests/test_basic_twoword_classifier.py (hunk 7 above)
Comment (evvahu) [reply tests/test_basic_twoword_classifier.py:18]: should be okay now because I changed the mistake above

**Review by Blubberli — APPROVED** [review]
